### DISCUSSION TITLE
x-for is creating an undefined element

### DISCUSSION THREAD

**Discussion by edwardexo** [alpinejs/alpine#3657] [7. Bugs (archived)] [answered]
I have the following JSON data:

```javascript
permissionCategories: [
    {
        "name": "Category 1",
        "permissions": [
             {
                 "code": "permission_code_1",
                 "permissions": [
                     {
                         "code": "sub_permission_code_1",
                     }
                  ],
              }
        ]
    },
    {
        "name": "Category 2",
        "permissions": [
            {
                 "code": "permission_code_2"
            }
          ]
    }
]
```

With the following code:

```javascript
<template x-for="permissionCategory in permissionCategories">
    <div>
        <template x-for="permission in permissionCategory.permissions">
            <div>
                main permission

                <template x-if="typeof permission.permissions !== 'undefined'" x-for="(subPermission, index) in permission.permissions" hidden>
                    <span>
                        ---- sub permission
                        <span x-text="index"></span><br>
                    </span>
                </template>
            </div>
        </template>
    </div>
</template>
```

**Description:**
The code is supposed to show all category permissions and their subpermissions (if such exist). As can be seen from my data, the first permission has one subpermission, which should get displayed. For whatever reason, two items get displayed. The first one is literally `undefined` (with no index), while the second item is correct.

**Expected behaviour:**
The subpermission to be displayed.
![image](https://github.com/alpinejs/alpine/assets/118666903/95d9225f-e753-4aa7-8c44-8fa1fb08c31e)

**Actual behaviour:**
In addition to the subpermission being displayed, an `undefined` item is also in the list.
![image](https://github.com/alpinejs/alpine/assets/118666903/8af01c16-43f0-46bb-ab8b-e9d54dad8a98)

**Comment by ekwoka** [comment, marked as answer]
You can't put x-if and x-for on the same template. It will make elements for the x-for loop and an elements from the x-if.

For what you are doing here, change that most inner x-for to

```
x-for="(subPermission, index) in permission.permissions ?? []"
```

and remove the x-if entirely.
